File: nearty-backend/app/Http/Controllers/api/OrderController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Order;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Validator;

class OrderController extends Controller
{
    // GET /api/orders/my  (pesanan milik pembeli yang login)
    public function my(Request $request)
    {
        $orders = Order::with('review')
            ->where('user_id', $request->user()->id)
            ->latest()
            ->get();

        return response()->json($orders);
    }

    // GET /api/orders/available  (pesanan pending yang belum diambil driver manapun)
    public function available(Request $request)
    {
        $orders = Order::whereNull('driver_id')
            ->where('status', 'pending')
            ->latest()
            ->get();

        return response()->json($orders);
    }

    // GET /api/orders/driver/active  (pesanan yang sedang dijalani driver)
    public function driverActive(Request $request)
    {
        $orders = Order::where('driver_id', $request->user()->id)
            ->whereIn('status', ['accepted', 'picked_up'])
            ->latest()
            ->get();

        return response()->json($orders);
    }

    // GET /api/orders/driver/history
    public function driverHistory(Request $request)
    {
        $orders = Order::with('review')
            ->where('driver_id', $request->user()->id)
            ->whereIn('status', ['completed', 'cancelled'])
            ->latest()
            ->get();

        return response()->json($orders);
    }

    // POST /api/orders
    public function store(Request $request)
    {
        $validator = Validator::make($request->all(), [
            'item_name' => ['required', 'string', 'max:255'],
            'pickup_location' => ['required', 'string', 'max:255'],
            'destination_location' => ['required', 'string', 'max:255'],
            'fee' => ['required', 'integer', 'min:0'],
            'payment_method' => ['nullable', 'string', 'max:50'],
            'pickup_lat' => ['nullable', 'numeric'],
            'pickup_lng' => ['nullable', 'numeric'],
            'dest_lat' => ['nullable', 'numeric'],
            'dest_lng' => ['nullable', 'numeric'],
        ]);

        if ($validator->fails()) {
            return response()->json(['message' => 'Gagal membuat pesanan', 'errors' => $validator->errors()], 422);
        }

        $order = Order::create([
            'user_id' => $request->user()->id,
            'item_name' => $request->item_name,
            'pickup_location' => $request->pickup_location,
            'destination_location' => $request->destination_location,
            'fee' => $request->fee,
            'payment_method' => $request->payment_method ?? 'Cash',
            'pickup_lat' => $request->pickup_lat,
            'pickup_lng' => $request->pickup_lng,
            'dest_lat' => $request->dest_lat,
            'dest_lng' => $request->dest_lng,
            'status' => 'pending',
        ]);

        return response()->json($order, 201);
    }

    // PUT /api/orders/{id}  (hanya pemilik pesanan, selama masih pending)
    public function update(Request $request, Order $order)
    {
        if ($order->user_id !== $request->user()->id) {
            return response()->json(['message' => 'Gagal memperbarui pesanan'], 403);
        }

        $order->update($request->only(['item_name', 'pickup_location', 'destination_location']));

        return response()->json($order);
    }

    // DELETE /api/orders/{id}  (pembeli membatalkan pesanan miliknya -> soft cancel)
    public function destroy(Request $request, Order $order)
    {
        if ($order->user_id !== $request->user()->id) {
            return response()->json(['message' => 'Gagal membatalkan pesanan'], 403);
        }

        $order->update(['status' => 'cancelled']);

        return response()->json(['message' => 'Pesanan dibatalkan']);
    }

    // POST /api/orders/{id}/accept  (driver mengambil pesanan)
    public function accept(Request $request, Order $order)
    {
        if ($order->status !== 'pending' || $order->driver_id !== null) {
            return response()->json(['message' => 'Gagal mengambil pesanan'], 409);
        }

        $order->update([
            'driver_id' => $request->user()->id,
            'status' => 'accepted',
        ]);

        return response()->json(['order' => $order->fresh()]);
    }

    // PUT /api/orders/{id}/status  (driver update status: picked_up / completed)
    public function updateStatus(Request $request, Order $order)
    {
        $validator = Validator::make($request->all(), [
            'status' => ['required', 'string', 'in:pending,accepted,picked_up,completed,cancelled'],
        ]);

        if ($validator->fails()) {
            return response()->json(['message' => 'Status tidak valid'], 422);
        }

        if ($order->driver_id !== $request->user()->id) {
            return response()->json(['message' => 'Gagal memperbarui status'], 403);
        }

        $order->update(['status' => $request->status]);

        return response()->json($order->fresh());
    }

    // POST /api/orders/{id}/review
    public function review(Request $request, Order $order)
    {
        $validator = Validator::make($request->all(), [
            'rating' => ['required', 'integer', 'min:1', 'max:5'],
            'comment' => ['nullable', 'string', 'max:1000'],
        ]);

        if ($validator->fails()) {
            return response()->json(['message' => 'Gagal mengirim ulasan'], 422);
        }

        if ($order->user_id !== $request->user()->id) {
            return response()->json(['message' => 'Gagal mengirim ulasan'], 403);
        }

        $review = $order->review()->updateOrCreate(
            ['order_id' => $order->id],
            ['rating' => $request->rating, 'comment' => $request->comment]
        );

        return response()->json($review, 201);
    }
}

File: lib/controllers/order_controller.dart
import 'package:get/get.dart';
import '../models/order_model.dart';
import 'package:flutter/material.dart';
import 'dart:math';
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:latlong2/latlong.dart';

class OrderController extends GetxController {
  var myOrders = <Order>[].obs;
  var availableOrders = <Order>[].obs;
  var driverOrders = <Order>[].obs;
  var driverHistoryOrders = <Order>[].obs;
  final String baseUrl = 'http://172.16.11.79:8000/api';

  @override
  void onInit() {
    super.onInit();
    fetchMyOrders();
    fetchAvailableOrders();
    fetchDriverActiveOrders();
    fetchDriverHistory();
  }

  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('token');
  }

  Future<void> fetchMyOrders() async {
    final token = await _getToken();
    if (token == null) return;

    try {
      final response = await http.get(
        Uri.parse('$baseUrl/orders/my'),
        headers: {'Authorization': 'Bearer $token', 'Accept': 'application/json'},
      );
      if (response.statusCode == 200) {
        List data = jsonDecode(response.body);
        myOrders.value = data.map((e) => Order.fromJson(e)).toList();
      }
    } catch (e) {
      print('Error fetching my orders: $e');
    }
  }

  Future<void> fetchAvailableOrders() async {
    final token = await _getToken();
    if (token == null) return;

    try {
      final response = await http.get(
        Uri.parse('$baseUrl/orders/available'),
        headers: {'Authorization': 'Bearer $token', 'Accept': 'application/json'},
      );
      if (response.statusCode == 200) {
        List data = jsonDecode(response.body);
        availableOrders.value = data.map((e) => Order.fromJson(e)).toList();
      }
    } catch (e) {
      print('Error fetching available orders: $e');
    }
  }

  Future<void> fetchDriverActiveOrders() async {
    final token = await _getToken();
    if (token == null) return;

    try {
      final response = await http.get(
        Uri.parse('$baseUrl/orders/driver/active'),
        headers: {'Authorization': 'Bearer $token', 'Accept': 'application/json'},
      );
      if (response.statusCode == 200) {
        List data = jsonDecode(response.body);
        driverOrders.value = data.map((e) => Order.fromJson(e)).toList();
      }
    } catch (e) {
      print('Error fetching driver active orders: $e');
    }
  }

  Future<void> fetchDriverHistory() async {
    final token = await _getToken();
    if (token == null) return;

    try {
      final response = await http.get(
        Uri.parse('$baseUrl/orders/driver/history'),
        headers: {'Authorization': 'Bearer $token', 'Accept': 'application/json'},
      );
      if (response.statusCode == 200) {
        List data = jsonDecode(response.body);
        driverHistoryOrders.value = data.map((e) => Order.fromJson(e)).toList();
      }
    } catch (e) {
      print('Error fetching driver history: $e');
    }
  }

  Future<void> createOrder(String itemName, String pickupLocation, String destinationLocation, String paymentMethod, {double? pickupLat, double? pickupLng, double? destLat, double? destLng}) async {
    final token = await _getToken();
    
    int calculatedFee = 5000 + Random().nextInt(15) * 1000; // default fallback

    if (pickupLat != null && pickupLng != null && destLat != null && destLng != null) {
      final Distance distance = const Distance();
      final double km = distance.as(LengthUnit.Kilometer, LatLng(pickupLat, pickupLng), LatLng(destLat, destLng));
      
      // Base fee 5000, 3000 per KM
      calculatedFee = 5000 + (km * 3000).round();
      // Round to nearest 500 for cleaner numbers
      calculatedFee = (calculatedFee / 500).ceil() * 500;
      
      if (calculatedFee < 5000) calculatedFee = 5000;
    }

    try {
      final response = await http.post(
        Uri.parse('$baseUrl/orders'),
        headers: {'Authorization': 'Bearer $token', 'Content-Type': 'application/json', 'Accept': 'application/json'},
        body: jsonEncode({
          'item_name': itemName,
          'pickup_location': pickupLocation,
          'destination_location': destinationLocation,
          'fee': calculatedFee,
          'payment_method': paymentMethod,
          'pickup_lat': pickupLat,
          'pickup_lng': pickupLng,
          'dest_lat': destLat,
          'dest_lng': destLng,
        }),
      );

      if (response.statusCode == 201) {
        fetchMyOrders();
        Get.snackbar('Sukses', 'Pesanan sedang dicarikan driver!', backgroundColor: Colors.green, colorText: Colors.white);
      } else {
        String errMsg = 'Gagal membuat pesanan';
        try {
          final data = jsonDecode(response.body);
          if (data['message'] != null) errMsg = data['message'];
        } catch (_) {}
        Get.snackbar('Error', errMsg, backgroundColor: Colors.red, colorText: Colors.white);
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e', backgroundColor: Colors.red, colorText: Colors.white);
    }
  }

  Future<void> updateOrder(int id, String itemName, String pickupLocation, String destinationLocation) async {
    final token = await _getToken();
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/orders/$id'),
        headers: {'Authorization': 'Bearer $token', 'Content-Type': 'application/json', 'Accept': 'application/json'},
        body: jsonEncode({
          'item_name': itemName,
          'pickup_location': pickupLocation,
          'destination_location': destinationLocation,
        }),
      );
      if (response.statusCode == 200) {
        fetchMyOrders();
        Get.snackbar('Sukses', 'Pesanan diperbarui', backgroundColor: Colors.green, colorText: Colors.white);
      } else {
        String errMsg = 'Gagal memperbarui pesanan';
        try {
          final data = jsonDecode(response.body);
          if (data['message'] != null) errMsg = data['message'];
        } catch (_) {}
        Get.snackbar('Error', errMsg, backgroundColor: Colors.red, colorText: Colors.white);
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e', backgroundColor: Colors.red, colorText: Colors.white);
    }
  }

  Future<void> deleteOrder(int id) async {
    final token = await _getToken();
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/orders/$id'),
        headers: {'Authorization': 'Bearer $token', 'Accept': 'application/json'},
      );
      if (response.statusCode == 200) {
        fetchMyOrders();
        Get.snackbar('Sukses', 'Pesanan dibatalkan', backgroundColor: Colors.green, colorText: Colors.white);
      } else {
        String errMsg = 'Gagal membatalkan pesanan';
        try {
          final data = jsonDecode(response.body);
          if (data['message'] != null) errMsg = data['message'];
        } catch (_) {}
        Get.snackbar('Error', errMsg, backgroundColor: Colors.red, colorText: Colors.white);
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e', backgroundColor: Colors.red, colorText: Colors.white);
    }
  }

  Future<void> acceptOrder(Order order) async {
    final token = await _getToken();
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/orders/${order.id}/accept'),
        headers: {'Authorization': 'Bearer $token', 'Accept': 'application/json'},
      );
      if (response.statusCode == 200) {
        fetchAvailableOrders();
        final updatedData = jsonDecode(response.body)['order'];
        driverOrders.add(Order.fromJson(updatedData));
        Get.snackbar('Sukses', 'Pesanan diambil!', backgroundColor: Colors.green, colorText: Colors.white);
      } else {
        String errMsg = 'Gagal mengambil pesanan';
        try {
          final data = jsonDecode(response.body);
          if (data['message'] != null) errMsg = data['message'];
        } catch (_) {}
        Get.snackbar('Error', errMsg, backgroundColor: Colors.red, colorText: Colors.white);
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e', backgroundColor: Colors.red, colorText: Colors.white);
    }
  }

  Future<void> pickupOrder(Order order) async {
    final token = await _getToken();
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/orders/${order.id}/status'),
        headers: {'Authorization': 'Bearer $token', 'Content-Type': 'application/json', 'Accept': 'application/json'},
        body: jsonEncode({'status': 'picked_up'}),
      );
      if (response.statusCode == 200) {
        final idx = driverOrders.indexWhere((o) => o.id == order.id);
        if (idx != -1) {
          driverOrders[idx].status = 'picked_up';
          driverOrders.refresh();
        }
        Get.snackbar('Barang Diambil! 📦', 'Sekarang antar ke tujuan ya!',
            backgroundColor: const Color(0xFF4F46E5), colorText: Colors.white);
      } else {
        String errMsg = 'Gagal memperbarui status';
        try {
          final data = jsonDecode(response.body);
          if (data['message'] != null) errMsg = data['message'];
        } catch (_) {}
        Get.snackbar('Error', errMsg, backgroundColor: Colors.red, colorText: Colors.white);
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e', backgroundColor: Colors.red, colorText: Colors.white);
    }
  }

  Future<void> completeOrder(Order order) async {
    final token = await _getToken();
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/orders/${order.id}/status'),
        headers: {'Authorization': 'Bearer $token', 'Content-Type': 'application/json', 'Accept': 'application/json'},
        body: jsonEncode({'status': 'completed'}),
      );
      if (response.statusCode == 200) {
        driverOrders.removeWhere((o) => o.id == order.id);
        fetchDriverHistory();
        Get.snackbar('Pesanan Selesai! 🎉', 'Kerja bagus! Lanjut cari pesanan lagi.',
            backgroundColor: const Color(0xFF10B981), colorText: Colors.white);
      } else {
        String errMsg = 'Gagal menyelesaikan pesanan';
        try {
          final data = jsonDecode(response.body);
          if (data['message'] != null) errMsg = data['message'];
        } catch (_) {}
        Get.snackbar('Error', errMsg, backgroundColor: Colors.red, colorText: Colors.white);
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e', backgroundColor: Colors.red, colorText: Colors.white);
    }
  }

  Future<void> submitReview(int orderId, int rating, String comment) async {
    final token = await _getToken();
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/orders/$orderId/review'),
        headers: {'Authorization': 'Bearer $token', 'Content-Type': 'application/json', 'Accept': 'application/json'},
        body: jsonEncode({
          'rating': rating,
          'comment': comment,
        }),
      );
      if (response.statusCode == 201) {
        fetchMyOrders();
        Get.snackbar('Terima Kasih!', 'Ulasan kamu berhasil dikirim.', backgroundColor: Colors.green, colorText: Colors.white);
      } else {
        String errMsg = 'Gagal mengirim ulasan';
        try {
          final data = jsonDecode(response.body);
          if (data['message'] != null) errMsg = data['message'];
        } catch (_) {}
        Get.snackbar('Error', errMsg, backgroundColor: Colors.red, colorText: Colors.white);
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e', backgroundColor: Colors.red, colorText: Colors.white);
    }
  }
}
